### PR TITLE
Added conditional rendering of keyboard controls

### LINKED ISSUE #5047: When using Talkback, multiple swipes are required to navigate between messages in the transcript.
### Is it an issue related to Adaptive Cards?

No.

### What is the PWD impact?

Screen reader users will be impacted as the users get confused and need to swipe many times while navigating. The screen reader users efforts will be increased.

### What browsers and screen readers do this issue affect?

Android: Chrome with TalkBack

### Are there any code-based customization done to Web Chat?

No, I am using Web Chat without any customizations except "styleOptions".

### What version of Web Chat are you using?

Latest production

### Which area does this issue affect?

Transcript navigation

### What is the public URL for the website?

_No response_

### How to reproduce the issue?

1. With an Android device, navigate to https://compulim.github.io/webchat-loader/.
2. Select the preset "[Public] MockBot" and start a webchat.
3. Send at least two messages.
4. Enable TalkBack if not done already.
5. Tap the first message to focus it. TalkBack should read out its content.
6. With the first message in focus, swipe right to move focus to the next message.

### What do you expect?

Swiping right once should move focus to the next message.

### What actually happened?

Moving focus to the next message required five swipes.

1. After the first swipe, the timestamp for the first message was read out.
2. After the second swipe, nothing was read out.
3. After the third swipe, nothing was read out.
4. After the fourth swipe, "Bot said colon" was read out.
5. After the fifth swipe, the content of the second message was read out.

### **Environment Details:**

- Application Name: Chat widget
- URL: https://ppe2dot1with5dot1serorg3.crm10.dynamics.com/
- App name: Customer Service admin centerC
- Device: Android Version 14.
- Organization name: ppe2dot1with5dot1serorg3
- Server version: 9.2.23053.00184
- Client version: 1.4.5733-2305.3
- #URL: [React App (windows.net)](https://inapp.blob.core.windows.net/tools/omniwidget/index.html)
- Browser: Chrome

### Do you have any screenshots or recordings to repro the issue?

https://github.com/microsoft/BotFramework-WebChat/assets/120019572/eb5b853e-4a23-4101-a1c6-4b0cb0cf46a7

### Did you find any DOM elements that might have caused the issue?

After the second swipe when nothing is read out, I suspect TalkBack is reading the second FocusRedirector in the FocusTrap wrapping the first message. https://github.com/microsoft/BotFramework-WebChat/blob/58ec6cd3fcd19a8b070f9afad4ae05e6ab6881a4/packages/component/src/Transcript/FocusTrap.tsx#L56C1-L57C1

After the third swipe when nothing is read out, I suspect TalkBack is reading the first FocusRedirector in the FocusTrap wrapping the second message. https://github.com/microsoft/BotFramework-WebChat/blob/58ec6cd3fcd19a8b070f9afad4ae05e6ab6881a4/packages/component/src/Transcript/FocusTrap.tsx#L52C1-L53C1

### MAS reference

MAS 1.3.2 – Meaningful Sequence

### WCAG reference

_No response_

### WAI-ARIA reference

_No response_

### Adaptive Card JSON

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by amirmursal**
> Fixes #5047

## Changelog Entry

Fixes [#5047](https://github.com/microsoft/BotFramework-WebChat/issues/5047). Added condition for `FocusRedirector` in PR [#5206](https://github.com/microsoft/BotFramework-WebChat/pull/5206), by [@amirmursal](https://github.com/amirmursal) 

## Description

- Talkback is focusing on some empty divs while navigating through chat messages. This issue has been addressed in this PR

-  [ ] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [ ] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [ ] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [x] Browser and platform compatibilities reviewed
-  [ ] CSS styles reviewed (minimal rules, no `z-index`)
-  [ ] Documents reviewed (docs, samples, live demo)
-  [ ] Internationalization reviewed (strings, unit formatting)
-  [ ] `package.json` and `package-lock.json` reviewed
-  [ ] Security reviewed (no data URIs, check for nonce leak)
-  [ ] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: packages/component/src/Transcript/FocusTrap.tsx [hunk 1]
```diff
@@ -49,11 +49,11 @@ const FocusTrap: FC<FocusTrapProps> = ({ children, onFocus, onLeave }) => {
 
   return (
     <Fragment>
-      <FocusRedirector onFocus={handleFirstSentinelFocus} />
+      {!android && <FocusRedirector onFocus={handleFirstSentinelFocus} />}
```
Comment (sounakpal) [inline packages/component/src/Transcript/FocusTrap.tsx:52]: How will it work in android in this case?

Comment (sounakpal) [reply packages/component/src/Transcript/FocusTrap.tsx:52]: and others where you put !android as a check

Comment (amirmursal) [reply packages/component/src/Transcript/FocusTrap.tsx:52]: This **FocusRedirector** component creating extra div with **tabindex="0"** to mange the focus state. When user on android devoice with talkback feature on it won't be creating those extra divs. I had tested this code with talkback feature enable and it is working as expected. When talk back is off the focus state is also getting managed in android devoice.

**Comment by OEvgeny** [comment]
What about android devices with keyboard, like tablets and laptops?

File: packages/component/src/BasicTranscript.tsx [hunk 2]
```diff
@@ -345,16 +345,16 @@ const InternalTranscript = forwardRef<HTMLDivElement, InternalTranscriptProps>(
...
-                // Add subpixel tolerance
-                .find(([, element]) => {
-                  // "getClientRects()" is not returning an array, thus, it is not destructurable.
-                  // eslint-disable-next-line prefer-destructuring
-                  const elementClientRect = element.getClientRects()[0];
-
-                  // If the activity is not attached to DOM tree, we should not count it as "bottommost visible activity", as it is not visible.
-                  return elementClientRect && elementClientRect.bottom < scrollableClientBottom + 1;
-                })
+              .reverse()
```
Comment (OEvgeny) [inline packages/component/src/BasicTranscript.tsx:348]: nit: remove unwanted indentation changes

File: packages/component/src/BasicTranscript.tsx [hunk 3]
```diff
@@ -799,10 +799,10 @@ const useScroller = (activityElementMapRef: MutableRefObject<ActivityElementMap>
 
                 values.push(
                   nthUnacknowledgedActivityOffsetTop +
-                    nthUnacknowledgedActivityBoundingBoxElement.offsetHeight -
-                    offsetHeight -
-                    scrollTop +
-                    patchedAutoScrollSnapOnActivityOffset
+                  nthUnacknowledgedActivityBoundingBoxElement.offsetHeight -
```
Comment (OEvgeny) [inline packages/component/src/BasicTranscript.tsx:802]: nit: remove unwanted indentation changes

File: packages/component/src/BasicTranscript.tsx [hunk 4]
```diff
@@ -817,9 +817,9 @@ const useScroller = (activityElementMapRef: MutableRefObject<ActivityElementMap>
 
               values.push(
                 firstUnacknowledgedActivityOffsetTop -
-                  scrollTop -
-                  offsetHeight * (1 - patchedAutoScrollSnapOnPage) +
-                  patchedAutoScrollSnapOnPageOffset
+                scrollTop -
```
Comment (OEvgeny) [inline packages/component/src/BasicTranscript.tsx:820]: nit: remove unwanted indentation changes

**Review by OEvgeny — COMMENTED** [review]
@amirmursal while I understand, this would be an easy fix for android phones with talkback, there are a lot of android devices which could benefit from having the keyboard navigation. We should find a way to make our focus work consistently across platforms.

Perhaps you could try to run some android browser through Windows subsystem for Android to verify on the Android with keyboard type of devices in addition to phone

File: packages/component/src/BasicTranscript.tsx (hunk 2 above)
Comment (amirmursal) [reply packages/component/src/BasicTranscript.tsx]: fixed

File: packages/component/src/BasicTranscript.tsx (hunk 3 above)
Comment (amirmursal) [reply packages/component/src/BasicTranscript.tsx]: fixed

File: packages/component/src/BasicTranscript.tsx (hunk 4 above)
Comment (amirmursal) [reply packages/component/src/BasicTranscript.tsx]: fixed

File: packages/component/src/BasicTranscript.tsx [hunk 5]
```diff
@@ -862,4 +862,4 @@ BasicTranscript.propTypes = {
   className: PropTypes.string
 };
 
-export default memo(BasicTranscript);
+export default memo(BasicTranscript);
```
Comment (OEvgeny) [inline packages/component/src/BasicTranscript.tsx:865]: nit: remove unwanted change

Comment (amirmursal) [reply packages/component/src/BasicTranscript.tsx]: fixed

**Comment by OEvgeny** [comment]
> @amirmursal while I understand, this would be an easy fix for android phones with talkback, there are a lot of android devices which could benefit from having the keyboard navigation. We should find a way to make our focus work consistently across platforms.
> 
> Perhaps you could try to run some android browser through Windows subsystem for Android to verify on the Android with keyboard type of devices in addition to phone

@amirmursal this is not addressed still

**Pull request closed without merging**